drivers/platform/mbed/mbed_gpio.cpp

```diff
@@ -0,0 +1,308 @@
+ */
+int32_t mbed_gpio_direction_output(struct gpio_desc *desc, uint8_t value)
+{
+	DigitalOut *gpio_output;    	// pointer to gpio output object
+	struct mbed_gpio_desc
+		*gpio_desc_extra;   // pointer to gpio desc extra parameters
+
+	if (!desc|| !desc->extra)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+	if (gpio_desc_extra->gpio_pin == NULL) {
+		/* Configure and instantiate GPIO pin as output */
+		gpio_output = new DigitalOut((PinName)desc->number);
+		if (!gpio_output)
+			return -ENOMEM;
+
+		gpio_desc_extra->gpio_pin = (struct mbed_gpio_desc *)gpio_output;
+		gpio_desc_extra->direction = GPIO_OUT;
+
+		/* Set the GPIO value */
+		if (gpio_set_value(desc, value) != SUCCESS) {
+			free(gpio_output);
+			return -EIO;
+		}
+
+		return SUCCESS;
+	}
+
+	return -EINVAL;
+}
+
+/**
+ * @brief Get the direction of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @param direction[in,out] - The GPIO direction.
+ *                    Example: GPIO_OUT
+ *                             GPIO_IN
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get_direction(struct gpio_desc *desc, uint8_t *direction)
+{
+	mbed_gpio_desc *gpio_desc_extra;     // pointer to gpio desc extra parameters
+
+	if (!desc || !desc->extra || !direction)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+
+	if (gpio_desc_extra->gpio_pin)
+		*direction = gpio_desc_extra->direction;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the value of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @param value[in] - The GPIO value.
+ *                Example: GPIO_HIGH
+ *                         GPIO_LOW
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_set_value(struct gpio_desc *desc, uint8_t value)
+{
+	DigitalOut *gpio_output; 		// pointer to gpio output object
+	mbed_gpio_desc *gpio_desc_extra;    // pointer to gpio desc extra parameters
+
+	if (!desc|| !desc->extra)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+
+	if (gpio_desc_extra->gpio_pin) {
+		gpio_output = (DigitalOut *)((struct mbed_gpio_desc *)desc->extra)->gpio_pin;
+		gpio_output->write(value);
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the value of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @param value[in,out] - The GPIO value.
+ *                Example: GPIO_HIGH
+ *                         GPIO_LOW
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get_value(struct gpio_desc *desc, uint8_t *value)
+{
+	DigitalIn *gpio_input;    	// pointer to gpio input object
+	int32_t ret;
+
+	if (!desc || !desc->extra || !value)
+		return -EINVAL;
+
+	gpio_input = (DigitalIn *)((struct mbed_gpio_desc *)desc->extra)->gpio_pin;
+	*value = (uint8_t)gpio_input->read();
+	ret = gpio_input->is_connected() ? SUCCESS : -EIO;
```

Comment:
Question: should this be done before trying to read the GPIO value? And should we do this in every function before this?

Reply:
Good point. I think it doesn't make any harm to use this check for every function as it just checks if pin is mapped to valid GPIO and not NC.